### PR TITLE
Panic on closed channel

### PR DESCRIPTION
**Pull request by sfleiter**
I had a look on https://github.com/tg44/heptapod/issues/15 since I was "lucky" and can reproduce this more than 95% of runs.

I had to restructure the code quite a bit but avoided to change any behavior. 
Please report if you see anything.
For me this runs now successfully every time.

Most important part is to **not** close any channel.
In go you only have to close channels if you want to use that as a mechanism to tell the receiver no more values are coming, see f.e. https://go.dev/tour/concurrency/4

For everything else please look through the individual commits, that will probably be easier to understand the refactoring.

### CODE REVIEW

File: pkg/walker/walker.go [hunk 1]
```diff
@@ -27,60 +28,54 @@ type WalkJob struct {
...
 
-	res := []string{}
-	//wait loop
-	<-start
-	for e := range end {
-		more := false
+	globalResult := []string{}
+	expectedResultCount := initialJobCount
+	maxId := 0
+	for expectedResultCount > 0 {
```
Comment (tg44) [inline pkg/walker/walker.go:51]: Don't we need to add a `|| len(spawn) > 0` also?

If `expectedResultCount` is 1, par is 1, and we get a new job to the `jobQueue`, when we get out the `singleResult`, we decrease the `expectedResultCount` to 0 and add a new element to the `spawn` and the for will stop while we still have an unprocessed element in `spawn`.

**Review by tg44 — CHANGES_REQUESTED** [review]
I think this is smart, only one question below;

File: pkg/walker/walker.go (hunk 1 above)
Comment (sfleiter) [reply pkg/walker/walker.go:51]: Thanks a lot for giving it your additional thoughts!

Based on this I moved to counting running jobs which should be more robust and is easier to understand.
Update coming 😄

Comment (sfleiter) [reply pkg/walker/walker.go]: Done, what do you think?

**Pull request merged**